packaging/material/packaging_python.md

```diff
@@ -0,0 +1,180 @@
+└── src/
+    └── solver/
+        ├── __init__.py
+        └── solver.py
+
+- The file `__init__.py` is required to import the `src/` as a package. This file is mostly empty
+- `solver.py` will contain the code developed by the user
+
+---
+
+## File structure for code packaging 2/2
+
+- Once the file structuring is complete, the repository will look like:
+
+package_name/
+├── LICENSE
+├── setup.py
+├── setup.cfg
+├── README.md
+├── src/
+│   └── solver/
+│       ├── __init__.py
+│       └── solver.py
+└── tests/
+
+- We will now look at individual components in greater detail
+
+---
+
+## setup.py
+
+- We will use [setuptools](https://pypi.org/project/setuptools/) to define the installation recipe for the package
+- `setuptools` can be configured using a config script called `setup.cfg` and the build recipe can be defined in a build script called `setup.py`
+    - `setup.cfg` consists of details like name of the package, version of package, files to include in the package.
+    - `setup.py` consists of configuration information as well as information on dependencies and details to pass on to PyPI after publishing
+
+---
+
+## LICENSE
+
+- This file provides details on which License is the package distributed under
+- Choosing the correct license is often a critical step in open-source software development
+- More on licenses in a future chapter of this course
+
+---
+
+## README.md
+
+- A `README` file typically involves detailed description about the package and also the installation procedure#
+- The `setuptools` configuration loads the `README.md` file to get the `long_description`, hence it is mandatory to include this file
+- Newer versions of `setuptools` also include this file automatically
+
+---
+
+## Creating distribution archives 1/2
+
+- We need to convert the folder sturcture that we have into an archive file system which the end user can download and install on their system
+- To do this we will use the package builder [build](https://pypa-build.readthedocs.io/en/stable/index.html) which was introduced in [PEP 517](https://www.python.org/dev/peps/pep-0517/)
+- These archive files are finally uploaded to PyPI and can be installed by pip
+- The disctribution archives are generated by running the command `python3 setup.py sdist` in the `package_name/` repository
+- The following directory and files will be generated in the repository:
+
+```python
+dist/
+  package_name_YOUR_USERNAME_HERE-0.0.1-py3-none-any.whl
+  package_name_YOUR_USERNAME_HERE-0.0.1.tar.gz
+```
+
+---
+
+## Uploading the distribution archives 1/2
+
+- Once the distribution archives are created, we want to upload it to the Python Package Index
+- We will use [Twine](https://twine.readthedocs.io/en/latest/) to upload our package to PyPI
+    - Twine is widely used as it securely authenticates the user to PyPI over HTTPS using a verified connection
+    - Its predecessor `python setup.py upload` required careful configuration of the user system and Python version
```

Comment:
What is Twine actually doing and why is it necessary? It is unclear to me why one needs a special tool for this.